### PR TITLE
stdlib: add unavailable methods on string indices (new indexing model migration)

### PR DESCRIPTION
**Pull request by gribozavr**
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@amartini51 @natecook1000 Could you review the messages for the unavailable methods?

File: stdlib/public/core/UnavailableStringAPIs.swift.gyb [hunk 1]
```diff
@@ -109,3 +109,41 @@ ${stringSubscriptComment}
...
   }
 }
+
+% for View in ['Character', 'UnicodeScalar', 'UTF16', 'UTF8']:
+%   Index = 'String.%sView.Index' % View
+%   Distance = 'String.%sView.Index' % View
+extension ${Index} {
+  @available(
+    *, unavailable,
+    message: "To get the next index call 'index(after:)' on the instance of the ${View} string view that produced the index")
```
Comment (natecook1000) [inline stdlib/public/core/UnavailableStringAPIs.swift.gyb]: These will end up saying something like:

    To get the next index call 'index(after:)' on the instance of the UTF8 string view that produced the index

But `Character`...`UTF8` aren't the right type and won't point them to the right place—I think I'd prefer to say:

    To get the next index call 'index(after:)' on the ${View}View instance that produced the index

or, more briefly

    To get the next index call 'index(after:)' on the related ${View}View instance

File: stdlib/public/core/UnavailableStringAPIs.swift.gyb [hunk 2]
```diff
@@ -109,3 +109,41 @@ ${stringSubscriptComment}
...
+  public func successor() -> ${Index} {
+    Builtin.unreachable()
+  }
+  @available(
+    *, unavailable,
+    message: "To get the previous index call 'index(after:)' on the instance of the ${View} string view that produced the index")
```
Comment (natecook1000) [inline stdlib/public/core/UnavailableStringAPIs.swift.gyb]: Should be `index(before:)`—also, skip this for `UTF8View.Index` or have the message say that it isn't bidirectional?

**Comment by gribozavr** [comment]
@natecook1000 Thanks, pushed an update!  How does it look?

**Comment by natecook1000** [comment]
Looks great!

**Comment by gribozavr** [comment]
@swift-ci Please test and merge

**Comment by gribozavr** [comment]
Thanks @natecook1000!

**Comment by gribozavr** [comment]
CI failure is unrelated.

**Pull request merged**